File: 백준/Silver/4889. 안정적인 문자열/안정적인 문자열.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        for (int t = 1;;t++) {
            String word = br.readLine();
            if (word.charAt(0) == '-') break;

            int count = 0;
            int stack = 0;

            for (char ch: word.toCharArray()) {
                switch (ch) {
                    case '{':
                        stack++;
                        break;
                    case '}':
                    {
                        if (stack == 0) {
                            count++;
                            stack++;
                        } else {
                            stack--;
                        }
                    }
                }
            }

            count += stack / 2;
            sb.append(t).append(". ").append(count).append("\n");
        }

        System.out.print(sb);
    }
}

File: 백준/Silver/4889. 안정적인 문자열/안정적인 문자열.py
import sys
input = sys.stdin.readline

def solution(T, word):
    count = 0
    stack = 0

    for ch in word:
        if ch == '{':
            stack += 1

        else:
            if stack == 0:
                count += 1
                stack += 1
            else:
                stack -= 1
    count += stack // 2
    print(f'{T}. {count}')


T = 1
while (word := input().strip())[0] != '-':
    solution(T, word)
    T += 1